### PR TITLE
HTTP2: Implement fairness when deferred processing streams.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

Commit Message: HTTP2: Implement fairness when deferred processing streams.
Additional Description: We will now round robin prioritizing streams that haven't written to the connection recently to process deferred data when we go from network connection backed up -> network connection low watermark. Still need to implement chunking.
Risk Level: low (off by default due to release guard)
Testing: unit / integration
Docs Changes: NA (until more fleshed out)
Release Notes: NA (until feature fully implemented)
Platform Specific Features: NA
Related Issue #19127

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @RyanTheOptimist 

Mind taking a look since Yan and Alyssa are out? Thanks!

File: source/common/http/http2/codec_impl.h [hunk 1]
```diff
@@ -169,8 +169,9 @@ class ConnectionImpl : public virtual Connection,
     }
   }
   void onUnderlyingConnectionBelowWriteBufferLowWatermark() override {
-    for (auto& stream : active_streams_) {
-      stream->runLowWatermarkCallbacks();
+    // Notify the streams based on least recently encoding to the connection.
+    for (auto it = active_streams_.rbegin(); it != active_streams_.rend(); it++) {
```
Comment (RyanTheOptimist) [inline source/common/http/http2/codec_impl.h:173]: nit: ++i, as per the style guide. (Not that it terribly matters here)

File: source/common/http/http2/codec_impl.h [hunk 2]
```diff
@@ -169,8 +169,9 @@ class ConnectionImpl : public virtual Connection,
...
+      (*it)->runLowWatermarkCallbacks();
```
Comment (RyanTheOptimist) [inline source/common/http/http2/codec_impl.h:174]: I notice that the code in  `updateActiveStreamsOnEncode()` is guarded by `stream.defer_processing_backedup_streams_` but the change here to process the list in reverse order is not. Can you say more about this?

File: test/common/http/http2/codec_impl_test.cc [hunk 3]
```diff
@@ -3643,6 +3652,69 @@ TEST_P(Http2CodecImplTest, CanHandleMultipleBufferedDataProcessingOnAStream) {
   }
 }
 
+TEST_P(Http2CodecImplTest, ShouldTrackWhichStreamLeastRecentlyEncodedIfDeferProcessing) {
```
Comment (RyanTheOptimist) [inline test/common/http/http2/codec_impl_test.cc:3655]: I'm not sure I understand where in this test we're testing the new behavior. I'm sure it does so, but there's a lot of code and none of the comments seem to make it clear. Where is the key EXPECT?

File: test/integration/buffer_accounting_integration_test.cc [hunk 4]
```diff
@@ -1089,6 +1128,232 @@ TEST_P(Http2DeferredProcessingIntegrationTest,
...
+  // processing.
+  writeToRequests(request_response_pairs, {2500, 2500, 2500, 2500}, false);
+
+  // Verify codec received but is buffered as we're still read disabled.
+  test_server_->waitForCounterGe("http.config_test.downstream_cx_rx_bytes_total", 20000);
+  test_server_->waitForCounterEq("http.config_test.downstream_flow_control_resumed_reading_total",
+                                 0);
+
+  // Track the order of when streams invoke decodeData to understand the
+  // ordering in deferred processing.
```
Comment (RyanTheOptimist) [inline test/integration/buffer_accounting_integration_test.cc:1165]: These comments are great, by the way!

**Review by RyanTheOptimist — COMMENTED** [review]
I think it would be a good idea to mention "HTTP/2" in the PR description since this does not affect HTTP/1 or HTTP/3. Also, does the change from nghttp2 to oghttp2 affect this at all, or is that code at a lower layer than this PR works with?

Looks like there is a build error in test/integration/buffer_accounting_integration_test.cc, btw.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review, will update to mention HTTP2.

This is orthogonal  to the underlying codec as this is at a higher level

**Review by RyanTheOptimist — APPROVED** [review]
LGTM

**Comment by RyanTheOptimist** [comment]
Looks like it needs a main merge and CI is broken?

**Comment by RyanTheOptimist** [comment]
@mattklein123 for non-Google review

**Review by mattklein123 — APPROVED** [review]
LGTM from a high level skim. Thank you!

**Pull request merged**